presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughUnion.java

```diff
@@ -59,33 +60,38 @@ public Result apply(ProjectNode parent, Captures captures, Context context)
 
         for (int i = 0; i < source.getSources().size(); i++) {
-            Map<Symbol, SymbolReference> outputToInput = Maps.transformValues(source.sourceSymbolMap(i), Symbol::toSymbolReference);   // Map: output of union -> input of this source to the union
+            Map<VariableReferenceExpression, SymbolReference> outputToInput = Maps.transformValues(source.sourceVariableMap(i), variable -> new SymbolReference(variable.getName()));   // Map: output of union -> input of this source to the union
             Assignments.Builder assignments = Assignments.builder(); // assignments for the new ProjectNode
 
             // mapping from current ProjectNode to new ProjectNode, used to identify the output layout
-            Map<Symbol, Symbol> projectSymbolMapping = new HashMap<>();
+            Map<VariableReferenceExpression, VariableReferenceExpression> projectVariableMapping = new HashMap<>();
 
             // Translate the assignments in the ProjectNode using symbols of the source of the UnionNode
-            for (Map.Entry<Symbol, Expression> entry : parent.getAssignments().entrySet()) {
-                Expression translatedExpression = inlineSymbols(outputToInput, entry.getValue());
-                Type type = context.getSymbolAllocator().getTypes().get(entry.getKey());
-                Symbol symbol = context.getSymbolAllocator().newSymbol(translatedExpression, type);
-                assignments.put(symbol, translatedExpression);
-                projectSymbolMapping.put(entry.getKey(), symbol);
+            for (Map.Entry<VariableReferenceExpression, Expression> entry : parent.getAssignments().entrySet()) {
+                Expression translatedExpression = inlineVariables(outputToInput, entry.getValue(), context.getSymbolAllocator().getTypes());
+                Type type = entry.getKey().getType();
+                VariableReferenceExpression variable = context.getSymbolAllocator().newVariable(translatedExpression, type);
+                assignments.put(variable, translatedExpression);
+                projectVariableMapping.put(new VariableReferenceExpression(entry.getKey().getName(), type), variable);
             }
             outputSources.add(new ProjectNode(context.getIdAllocator().getNextId(), source.getSources().get(i), assignments.build()));
-            outputLayout.forEach(symbol -> mappings.put(symbol, projectSymbolMapping.get(symbol)));
+            outputLayout.forEach(variable -> mappings.put(variable, projectVariableMapping.get(variable)));
         }
 
-        return Result.ofPlanNode(new UnionNode(parent.getId(), outputSources.build(), mappings.build(), ImmutableList.copyOf(mappings.build().keySet())));
+        return Result.ofPlanNode(new UnionNode(parent.getId(), outputSources.build(), mappings.build()));
+    }
+
+    private static VariableReferenceExpression getWithMatchingSymbol(Map<VariableReferenceExpression, VariableReferenceExpression> variableMapping, Symbol symbol)
```

Comment:
Not used anymore?

Reply:
If you can point out from which commit this is no longer used, that would be very helpful. 😂Otherwise I think I'll just remove them in follow up diffs...

Reply:
Given the size of the PR, follow up clean up diff/PR sounds great :)